### Create day datasets of 500 samples max for the small tweet dataset

In [5]:
pip install whatthelang

^C
ERROR: Operation cancelled by user
Note: you may need to restart the kernel to use updated packages.


In [9]:
import numpy as np
import pandas as pd
from whatthelang import WhatTheLang

from tqdm import tqdm

import datetime
import os
from IPython.display import FileLink

In [7]:
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

/kaggle/input/bitcoin-tweets/Bitcoin_tweets.csv


In [11]:
df = pd.read_csv('/kaggle/input/bitcoin-tweets/Bitcoin_tweets.csv', engine='python')

In [12]:
df.head()

user_name    user_location  \
0                             DeSota Wilson      Atlanta, GA   
1                                  CryptoND              NaN   
2                                 Tdlmatias  London, England   
3                      Crypto is the future              NaN   
4  Alex Kirchmaier 🇦🇹🇸🇪 #FactsSuperspreader           Europa   

                                    user_description         user_created  \
0  Biz Consultant, real estate, fintech, startups...  2009-04-26 20:05:09   
1  😎 BITCOINLIVE is a Dutch platform aimed at inf...  2019-10-17 20:12:10   
2  IM Academy : The best #forex, #SelfEducation, ...  2014-11-10 10:50:37   
3  I will post a lot of buying signals for BTC tr...  2019-09-28 16:48:12   
4  Co-founder @RENJERJerky | Forbes 30Under30 | I...  2016-02-03 13:15:55   

  user_followers user_friends user_favourites user_verified  \
0         8534.0         7605            4838         False   
1         6769.0         1532           25483         False   
2          128.0          332             924         False   
3          625.0          129              14         False   
4         1249.0         1472           10482         False   

                  date                                               text  \
0  2021-02-10 23:59:04  Blue Ridge Bank shares halted by NYSE after #b...   
1  2021-02-10 23:58:48  😎 Today, that's this #Thursday, we will do a "...   
2  2021-02-10 23:54:48  Guys evening, I have read this article about B...   
3  2021-02-10 23:54:33  $BTC A big chance in a billion! Price: \487264...   
4  2021-02-10 23:54:06  This network is secured by 9 508 nodes as of t...   

                                    hashtags               source is_retweet  
0                                ['bitcoin']      Twitter Web App      False  
1  ['Thursday', 'Btc', 'wallet', 'security']  Twitter for Android      False  
2                                        NaN      Twitter Web App      False  
3         ['Bitcoin', 'FX', 'BTC', 'crypto']              dlvr.it      False  
4                                    ['BTC']      Twitter Web App      False

In [13]:
df.size

57017493

In [16]:
df.drop(['user_location', 'user_description', 'user_favourites', 'user_created', 'user_followers', 'user_friends', 'user_verified', 'source', 'hashtags', 'is_retweet'], axis=1, inplace=True)

In [28]:
df.head()

user_name                 date  \
0                             DeSota Wilson  2021-02-10 23:59:04   
1                                  CryptoND  2021-02-10 23:58:48   
2                                 Tdlmatias  2021-02-10 23:54:48   
3                      Crypto is the future  2021-02-10 23:54:33   
4  Alex Kirchmaier 🇦🇹🇸🇪 #FactsSuperspreader  2021-02-10 23:54:06   

                                                text  
0  Blue Ridge Bank shares halted by NYSE after #b...  
1  😎 Today, that's this #Thursday, we will do a "...  
2  Guys evening, I have read this article about B...  
3  $BTC A big chance in a billion! Price: \487264...  
4  This network is secured by 9 508 nodes as of t...

In [29]:
df.tail()

user_name                 date  \
4385956           需要粉丝的进主页查询  2022-10-12 19:52:56   
4385957         🏹snipers 🏹 🚀  2022-10-12 19:52:46   
4385958      All Things Defi  2022-10-12 19:52:45   
4385959  Paul Barron Network  2022-10-12 19:52:36   
4385960           需要粉丝的进主页查询  2022-10-12 19:52:30   

                                                      text  
4385956  A good f#粉丝 #twitter粉丝  #twitter刷粉  #blockchai...  
4385957                            #btc pump in tbe way ✍️  
4385958  updated price for $EOS, $AVAX, $MATIC:\n\n#EOS...  
4385959  Can #google be the one?!! 🤔 #Coinbase #BTC #ET...  
4385960  vulgar,but you can #cyptocurrency #btc #eth #c...

In [26]:
df.dropna(subset=['date'], inplace=True)

In [30]:
df.isna().sum()

user_name    51
date          0
text          0
dtype: int64

In [39]:
df['date'] = pd.to_datetime(df['date'], errors='coerce')

In [40]:
# Create date column as time isn't required
def dt(x):
    t = pd.Timestamp(x)
    return pd.Timestamp.date(t)

df.date = df.date.apply(dt)

In [41]:
df.head()

user_name        date  \
0                             DeSota Wilson  2021-02-10   
1                                  CryptoND  2021-02-10   
2                                 Tdlmatias  2021-02-10   
3                      Crypto is the future  2021-02-10   
4  Alex Kirchmaier 🇦🇹🇸🇪 #FactsSuperspreader  2021-02-10   

                                                text  
0  Blue Ridge Bank shares halted by NYSE after #b...  
1  😎 Today, that's this #Thursday, we will do a "...  
2  Guys evening, I have read this article about B...  
3  $BTC A big chance in a billion! Price: \487264...  
4  This network is secured by 9 508 nodes as of t...

In [42]:
df = df.sort_values(by=['date'], ascending=True)

In [43]:
df.head()

user_name        date  \
693194                              FB- Xiomara Castañeda  0001-01-01   
64943   Can roam the world，It's nine to five againᵕᴗᵕ)...  0001-01-01   
180575                  • Learn n To Do d'Best...!!!!!!!!  0001-01-01   
137068   *Muhammad Yasir* hello stalker nice to tweet ...  0001-01-01   
697397          Pin bb : 26ea62f8 . Line : baliratih_bali  0001-01-01   

                       text  
693194      Twitter Web App  
64943       Twitter Web App  
180575      Twitter Web App  
137068  Twitter for Android  
697397      Twitter Web App

In [45]:
# Remove unnecessary data
df= df[df['date'] > datetime.date(2014, 1, 1)]

In [74]:
df[df['date'] == datetime.date(2021, 2, 10)]

user_name        date  \
0           DeSota Wilson  2021-02-10   
3525              PiRoyot  2021-02-10   
1180    The Crypto Geek⚡️  2021-02-10   
1179       Now on Bitcoin  2021-02-10   
1178       cryptotraderog  2021-02-10   
...                   ...         ...   
2318        Hartley louis  2021-02-10   
2350           CryptoBear  2021-02-10   
2352               BOCVIP  2021-02-10   
2351          CryptoParos  2021-02-10   
2349  MattMartProductions  2021-02-10   

                                                   text lang  
0     Blue Ridge Bank shares halted by NYSE after #b...   en  
3525  Market Id : BTC-AUD\nLast Price :60278.57\n24 ...   fr  
1180  Get ready for some Correction, Take the #profi...   en  
1179  🔼🔼 ₿1 = $44,724 (16:21 UTC)\n$BTC prices conti...   en  
1178  When I get into a short position on #Bitcoin ,...   en  
...                                                 ...  ...  
2318  @Lindsie_Rose Next Ticket Ready WEDNESDAY 10/2...   en  
2350  👀 new ath ??? 🙄📈\n\n$btc #BTC #Bitcoin https:/...   en  
2352  How to #Buy #Bitcoin in 2021. #Best ...\n\nhtt...   en  
2351  BULLSEYE ! 🎯🎯🎯\n\nAION / USDT (Before &amp; af...   en  
2349  @trapstar_231 @elonmusk How to buy #dogecoin #...   en  

[3526 rows x 4 columns]

In [44]:
df.tail()

user_name        date  \
4253041      Ethereum Yoda  2022-10-15   
4253042    Ojeniyi Ayobami  2022-10-15   
4253043                 JL  2022-10-15   
4253021    Data Metrics UG  2022-10-15   
4273457  deXit Price Bot 🤖  2022-10-15   

                                                      text  
4253041  #Ethereum price update: \n\n#ETH $1295.01 USD\...  
4253042  Exciting thing about @theZuriTeam is how the t...  
4253043  @BSCNews @BNBCHAIN @TiFiToken @BabyDogeCoin @b...  
4253021  💰 Market Cap. Swap 💰 What would one #Alchemix-...  
4273457  The current data for #deXit \n \nprice 0.0014 ...

In [46]:
df.head()

user_name        date  \
20395  ₿ullrun Opinions - Litecoin / BTC/ DOT/ ETH/ UDOO  2021-02-05   
20398                                          Mulon-Esk  2021-02-05   
20397                              𝘾𝙧𝙮𝙥𝙩𝙤+𝙎𝙥𝙤𝙧𝙩𝙨=𝘾𝙖𝙨𝙝 🏧₿  2021-02-05   
20396                                 Tiffany Martinez 🌴  2021-02-05   
20394                                  Flamezy Omo Abana  2021-02-05   

                                                    text  
20395  @daphantomman @Ilikedtolurk @novogratz I was t...  
20398  Some of these low caps in my portfolio are goi...  
20397  Would not be surprised to see a mini dump in #...  
20396  "#Technology, #Educational and #RewardsProgram...  
20394  Decided To Drop This Here!!!\nI Go Change Coun...

In [47]:
df.tail()

user_name        date  \
4253041      Ethereum Yoda  2022-10-15   
4253042    Ojeniyi Ayobami  2022-10-15   
4253043                 JL  2022-10-15   
4253021    Data Metrics UG  2022-10-15   
4273457  deXit Price Bot 🤖  2022-10-15   

                                                      text  
4253041  #Ethereum price update: \n\n#ETH $1295.01 USD\...  
4253042  Exciting thing about @theZuriTeam is how the t...  
4253043  @BSCNews @BNBCHAIN @TiFiToken @BabyDogeCoin @b...  
4253021  💰 Market Cap. Swap 💰 What would one #Alchemix-...  
4273457  The current data for #deXit \n \nprice 0.0014 ...

### Limit number of tweets #01

In [48]:
# limit number of tweets for each date to 500 due to computational limitations
df_grouped = df.groupby('date').head(500)

In [49]:
df_grouped.shape

(100633, 3)

In [50]:
df_grouped['date'].value_counts()

2021-02-05    500
2022-04-20    500
2022-03-21    500
2022-03-23    500
2022-03-24    500
             ... 
2021-08-21    500
2021-08-23    500
2021-04-17    333
2021-03-11    217
2022-03-22     83
Name: date, Length: 203, dtype: int64

In [51]:
df_grouped['date'].unique().size

203

In [52]:
os.makedirs('/kaggle/working/outputs')

In [53]:
# Create checkpoint file
df_grouped.to_csv('/kaggle/working/outputs/tweets_small_filtered.csv')

In [55]:
!zip -r file.zip './outputs'
FileLink(r'file.zip')

updating: outputs/ (stored 0%)
updating: outputs/tweets_small_filtered.csv (deflated 64%)


/kaggle/working/file.zip

### Remove null & non-english rows

In [56]:
df.dropna(subset=['user_name', 'date', 'text'], inplace=True)

In [57]:
df.size

13149405

Removed around 51 empty user names and 2000 ish empty dates

In [58]:
# Remove non-english text using WhatTheLang (https://github.com/indix/whatthelang)
wtl = WhatTheLang()
L = []

for row in tqdm(df['text']):
    if len(row) != 0:
        L.append(wtl.predict_lang(row))
    else:
        L.append(None)
        
df['lang'] = L
df.head()

100%|██████████| 4383135/4383135 [04:45<00:00, 15371.04it/s]


user_name        date  \
20395  ₿ullrun Opinions - Litecoin / BTC/ DOT/ ETH/ UDOO  2021-02-05   
20398                                          Mulon-Esk  2021-02-05   
20397                              𝘾𝙧𝙮𝙥𝙩𝙤+𝙎𝙥𝙤𝙧𝙩𝙨=𝘾𝙖𝙨𝙝 🏧₿  2021-02-05   
20396                                 Tiffany Martinez 🌴  2021-02-05   
20394                                  Flamezy Omo Abana  2021-02-05   

                                                    text lang  
20395  @daphantomman @Ilikedtolurk @novogratz I was t...   en  
20398  Some of these low caps in my portfolio are goi...   en  
20397  Would not be surprised to see a mini dump in #...   en  
20396  "#Technology, #Educational and #RewardsProgram...   en  
20394  Decided To Drop This Here!!!\nI Go Change Coun...   en

In [59]:
df_english = df[df['lang'] == 'en']

In [60]:
df_english.head()

user_name        date  \
20395  ₿ullrun Opinions - Litecoin / BTC/ DOT/ ETH/ UDOO  2021-02-05   
20398                                          Mulon-Esk  2021-02-05   
20397                              𝘾𝙧𝙮𝙥𝙩𝙤+𝙎𝙥𝙤𝙧𝙩𝙨=𝘾𝙖𝙨𝙝 🏧₿  2021-02-05   
20396                                 Tiffany Martinez 🌴  2021-02-05   
20394                                  Flamezy Omo Abana  2021-02-05   

                                                    text lang  
20395  @daphantomman @Ilikedtolurk @novogratz I was t...   en  
20398  Some of these low caps in my portfolio are goi...   en  
20397  Would not be surprised to see a mini dump in #...   en  
20396  "#Technology, #Educational and #RewardsProgram...   en  
20394  Decided To Drop This Here!!!\nI Go Change Coun...   en

In [61]:
df_english.drop(['lang'], axis=1, inplace=True)

/opt/conda/lib/python3.7/site-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


In [62]:
df_english.head()

user_name        date  \
20395  ₿ullrun Opinions - Litecoin / BTC/ DOT/ ETH/ UDOO  2021-02-05   
20398                                          Mulon-Esk  2021-02-05   
20397                              𝘾𝙧𝙮𝙥𝙩𝙤+𝙎𝙥𝙤𝙧𝙩𝙨=𝘾𝙖𝙨𝙝 🏧₿  2021-02-05   
20396                                 Tiffany Martinez 🌴  2021-02-05   
20394                                  Flamezy Omo Abana  2021-02-05   

                                                    text  
20395  @daphantomman @Ilikedtolurk @novogratz I was t...  
20398  Some of these low caps in my portfolio are goi...  
20397  Would not be surprised to see a mini dump in #...  
20396  "#Technology, #Educational and #RewardsProgram...  
20394  Decided To Drop This Here!!!\nI Go Change Coun...

### Limit number of tweets #02

In [63]:
df_grouped_2 = df_english.groupby('date').head(500)

In [64]:
# Create checkpoint file
df_grouped_2.to_csv('/kaggle/working/outputs/tweets_small_filtered_cleaned.csv')

In [65]:
!zip -r file.zip './outputs'
FileLink(r'file.zip')

updating: outputs/ (stored 0%)
updating: outputs/tweets_small_filtered.csv (deflated 64%)
  adding: outputs/tweets_small_filtered_cleaned.csv (deflated 64%)


/kaggle/working/file.zip

### Create day datasets (use if necessary)

In [66]:
df_days = [y for x, y in df_grouped_2.groupby('date')]

In [67]:
df_days[0]

user_name        date  \
20395  ₿ullrun Opinions - Litecoin / BTC/ DOT/ ETH/ UDOO  2021-02-05   
20398                                          Mulon-Esk  2021-02-05   
20397                              𝘾𝙧𝙮𝙥𝙩𝙤+𝙎𝙥𝙤𝙧𝙩𝙨=𝘾𝙖𝙨𝙝 🏧₿  2021-02-05   
20396                                 Tiffany Martinez 🌴  2021-02-05   
20394                                  Flamezy Omo Abana  2021-02-05   
...                                                  ...         ...   
19858                                      $Damxchange 💎  2021-02-05   
19859                                      $Damxchange 💎  2021-02-05   
19861                                   FinTech News Hub  2021-02-05   
19863                                       Crypto hour💎  2021-02-05   
19864                             Cryptocurrencies / EUR  2021-02-05   

                                                    text  
20395  @daphantomman @Ilikedtolurk @novogratz I was t...  
20398  Some of these low caps in my portfolio are goi...  
20397  Would not be surprised to see a mini dump in #...  
20396  "#Technology, #Educational and #RewardsProgram...  
20394  Decided To Drop This Here!!!\nI Go Change Coun...  
...                                                  ...  
19858  Bitcoin another solid 24 hours, over $24 Billi...  
19859  Bitcoin Cash another solid 24 hours, over $5 B...  
19861  Our featured article for today! Please don't f...  
19863  @GoingParabolic That’s my target to consider r...  
19864  🔄 Prices update in $EUR (1 hour):\n\n$BTC   - ...  

[500 rows x 3 columns]

In [68]:
df_days[len(df_days) - 1]

user_name        date  \
4244574          tictoknextblock  2022-10-15   
4244589              Ben Justman  2022-10-15   
4244573            Ethereum Yoda  2022-10-15   
4244575               Chainclock  2022-10-15   
4244588          Casper Willberg  2022-10-15   
...                          ...         ...   
4244060  Alex Mincu - mincua.eth  2022-10-15   
4244062             Crypto Vince  2022-10-15   
4244063          Cardano Tracker  2022-10-15   
4244064          Alan ₿ Watts ⚡️  2022-10-15   
4244065               James Bang  2022-10-15   

                                                      text  
4244574  🟩 Block 758794 ⛏ ⛏\n\n📅 15th Oct 2022, 15:16:5...  
4244589  GUYS WITH CHAIRS.\n#Bitcoin https://t.co/LSBBR...  
4244573  #Ethereum price update: \n\n#ETH $1286.99 USD\...  
4244575  Block 758,794 was just mined. It had 555 trans...  
4244588  Welcome, to the #Bitcoin Bar in the capital of...  
...                                                    ...  
4244060  For some people, that might be true. 😎😂\n\n#bi...  
4244062  The #Bitcoin #BTC Spot to Futures Volume Weigh...  
4244063  Current #Cardano price Alert: \n\n#ADA $0.367 ...  
4244064  Proper responses to common #Bitcoin related qu...  
4244065  Past financial crisis have been regional or wi...  

[500 rows x 3 columns]

In [69]:
len(df_days)

203

In [70]:
# Create individual CSVs for each day
for i in tqdm(df_days):
    filename = str(i.iloc[0]['date'])
    i.to_csv(f'/kaggle/working/outputs/{filename}.csv')

100%|██████████| 203/203 [00:01<00:00, 165.52it/s]


In [71]:
!zip -r file.zip './outputs'
FileLink(r'file.zip')

updating: outputs/ (stored 0%)
updating: outputs/tweets_small_filtered.csv (deflated 64%)
updating: outputs/tweets_small_filtered_cleaned.csv (deflated 64%)
  adding: outputs/2021-05-25.csv (deflated 63%)
  adding: outputs/2021-04-21.csv (deflated 64%)
  adding: outputs/2021-08-14.csv (deflated 62%)
  adding: outputs/2021-08-21.csv (deflated 62%)
  adding: outputs/2022-05-23.csv (deflated 61%)
  adding: outputs/2022-10-09.csv (deflated 70%)
  adding: outputs/2022-05-02.csv (deflated 66%)
  adding: outputs/2021-11-26.csv (deflated 59%)
  adding: outputs/2021-02-08.csv (deflated 64%)
  adding: outputs/2022-09-13.csv (deflated 66%)
  adding: outputs/2022-02-17.csv (deflated 64%)
  adding: outputs/2022-02-14.csv (deflated 63%)
  adding: outputs/2022-02-09.csv (deflated 71%)
  adding: outputs/2021-10-21.csv (deflated 60%)
  adding: outputs/2022-07-03.csv (deflated 63%)
  adding: outputs/2022-06-28.csv (deflated 66%)
  adding: outputs/2022-01-20.csv (deflated 65%)
  adding: outputs/2021-08-2

/kaggle/working/file.zip

Not all dates are available

### Check which dates are missing

In [3]:
from datetime import date, timedelta

start_date = date(2021, 2, 5) 
end_date = date(2022, 10, 15)
delta = end_date - start_date
days = []

for i in range(delta.days + 1):
    day = start_date + timedelta(days=i)
    days.append(day)

In [5]:
days[:10]

[datetime.date(2021, 2, 5),
 datetime.date(2021, 2, 6),
 datetime.date(2021, 2, 7),
 datetime.date(2021, 2, 8),
 datetime.date(2021, 2, 9),
 datetime.date(2021, 2, 10),
 datetime.date(2021, 2, 11),
 datetime.date(2021, 2, 12),
 datetime.date(2021, 2, 13),
 datetime.date(2021, 2, 14)]

In [10]:
import os
available_days = os.listdir('../../../data/Tweets/broken_small_sets_limited')

In [11]:
available_days = available_days[:-2]

In [12]:
available_days

['2021-02-05.csv',
 '2021-02-06.csv',
 '2021-02-07.csv',
 '2021-02-08.csv',
 '2021-02-09.csv',
 '2021-02-10.csv',
 '2021-02-13.csv',
 '2021-02-14.csv',
 '2021-02-15.csv',
 '2021-02-18.csv',
 '2021-02-19.csv',
 '2021-02-22.csv',
 '2021-02-28.csv',
 '2021-03-11.csv',
 '2021-03-12.csv',
 '2021-04-05.csv',
 '2021-04-06.csv',
 '2021-04-07.csv',
 '2021-04-08.csv',
 '2021-04-09.csv',
 '2021-04-10.csv',
 '2021-04-11.csv',
 '2021-04-12.csv',
 '2021-04-17.csv',
 '2021-04-18.csv',
 '2021-04-19.csv',
 '2021-04-20.csv',
 '2021-04-21.csv',
 '2021-04-22.csv',
 '2021-04-23.csv',
 '2021-04-24.csv',
 '2021-05-25.csv',
 '2021-05-26.csv',
 '2021-05-27.csv',
 '2021-05-28.csv',
 '2021-05-29.csv',
 '2021-06-20.csv',
 '2021-06-21.csv',
 '2021-06-22.csv',
 '2021-06-23.csv',
 '2021-07-02.csv',
 '2021-07-03.csv',
 '2021-07-04.csv',
 '2021-07-05.csv',
 '2021-07-16.csv',
 '2021-07-17.csv',
 '2021-07-18.csv',
 '2021-07-19.csv',
 '2021-07-20.csv',
 '2021-07-21.csv',
 '2021-07-22.csv',
 '2021-07-23.csv',
 '2021-07-24

In [15]:
len(days) - len(available_days)

415

415 days are missing... which is significant: more than a year worth of tweets are missing.

In [17]:
available_days = [day.replace('.csv', '') for day in available_days]
available_days

['2021-02-05',
 '2021-02-06',
 '2021-02-07',
 '2021-02-08',
 '2021-02-09',
 '2021-02-10',
 '2021-02-13',
 '2021-02-14',
 '2021-02-15',
 '2021-02-18',
 '2021-02-19',
 '2021-02-22',
 '2021-02-28',
 '2021-03-11',
 '2021-03-12',
 '2021-04-05',
 '2021-04-06',
 '2021-04-07',
 '2021-04-08',
 '2021-04-09',
 '2021-04-10',
 '2021-04-11',
 '2021-04-12',
 '2021-04-17',
 '2021-04-18',
 '2021-04-19',
 '2021-04-20',
 '2021-04-21',
 '2021-04-22',
 '2021-04-23',
 '2021-04-24',
 '2021-05-25',
 '2021-05-26',
 '2021-05-27',
 '2021-05-28',
 '2021-05-29',
 '2021-06-20',
 '2021-06-21',
 '2021-06-22',
 '2021-06-23',
 '2021-07-02',
 '2021-07-03',
 '2021-07-04',
 '2021-07-05',
 '2021-07-16',
 '2021-07-17',
 '2021-07-18',
 '2021-07-19',
 '2021-07-20',
 '2021-07-21',
 '2021-07-22',
 '2021-07-23',
 '2021-07-24',
 '2021-07-25',
 '2021-07-26',
 '2021-07-27',
 '2021-07-28',
 '2021-07-29',
 '2021-07-30',
 '2021-08-04',
 '2021-08-05',
 '2021-08-06',
 '2021-08-07',
 '2021-08-08',
 '2021-08-09',
 '2021-08-14',
 '2021-08-

In [18]:
days = [str(day) for day in days]
days

['2021-02-05',
 '2021-02-06',
 '2021-02-07',
 '2021-02-08',
 '2021-02-09',
 '2021-02-10',
 '2021-02-11',
 '2021-02-12',
 '2021-02-13',
 '2021-02-14',
 '2021-02-15',
 '2021-02-16',
 '2021-02-17',
 '2021-02-18',
 '2021-02-19',
 '2021-02-20',
 '2021-02-21',
 '2021-02-22',
 '2021-02-23',
 '2021-02-24',
 '2021-02-25',
 '2021-02-26',
 '2021-02-27',
 '2021-02-28',
 '2021-03-01',
 '2021-03-02',
 '2021-03-03',
 '2021-03-04',
 '2021-03-05',
 '2021-03-06',
 '2021-03-07',
 '2021-03-08',
 '2021-03-09',
 '2021-03-10',
 '2021-03-11',
 '2021-03-12',
 '2021-03-13',
 '2021-03-14',
 '2021-03-15',
 '2021-03-16',
 '2021-03-17',
 '2021-03-18',
 '2021-03-19',
 '2021-03-20',
 '2021-03-21',
 '2021-03-22',
 '2021-03-23',
 '2021-03-24',
 '2021-03-25',
 '2021-03-26',
 '2021-03-27',
 '2021-03-28',
 '2021-03-29',
 '2021-03-30',
 '2021-03-31',
 '2021-04-01',
 '2021-04-02',
 '2021-04-03',
 '2021-04-04',
 '2021-04-05',
 '2021-04-06',
 '2021-04-07',
 '2021-04-08',
 '2021-04-09',
 '2021-04-10',
 '2021-04-11',
 '2021-04-

In [19]:
missing_days = [day for day in days if day not in available_days]

In [20]:
missing_days

['2021-02-11',
 '2021-02-12',
 '2021-02-16',
 '2021-02-17',
 '2021-02-20',
 '2021-02-21',
 '2021-02-23',
 '2021-02-24',
 '2021-02-25',
 '2021-02-26',
 '2021-02-27',
 '2021-03-01',
 '2021-03-02',
 '2021-03-03',
 '2021-03-04',
 '2021-03-05',
 '2021-03-06',
 '2021-03-07',
 '2021-03-08',
 '2021-03-09',
 '2021-03-10',
 '2021-03-13',
 '2021-03-14',
 '2021-03-15',
 '2021-03-16',
 '2021-03-17',
 '2021-03-18',
 '2021-03-19',
 '2021-03-20',
 '2021-03-21',
 '2021-03-22',
 '2021-03-23',
 '2021-03-24',
 '2021-03-25',
 '2021-03-26',
 '2021-03-27',
 '2021-03-28',
 '2021-03-29',
 '2021-03-30',
 '2021-03-31',
 '2021-04-01',
 '2021-04-02',
 '2021-04-03',
 '2021-04-04',
 '2021-04-13',
 '2021-04-14',
 '2021-04-15',
 '2021-04-16',
 '2021-04-25',
 '2021-04-26',
 '2021-04-27',
 '2021-04-28',
 '2021-04-29',
 '2021-04-30',
 '2021-05-01',
 '2021-05-02',
 '2021-05-03',
 '2021-05-04',
 '2021-05-05',
 '2021-05-06',
 '2021-05-07',
 '2021-05-08',
 '2021-05-09',
 '2021-05-10',
 '2021-05-11',
 '2021-05-12',
 '2021-05-

The above dates are not available in small dataset